agent/internal/taskoutput/test_log.go

```diff
@@ -0,0 +1,330 @@
+
+	data, err := os.ReadFile(filepath.Join(h.dir, testLogSpecFilename))
+	if err != nil {
+		h.logger.Task().Warning(errors.Wrap(err, "reading test log spec; falling back to default spec"))
+		return
+	}
+	if err = yaml.Unmarshal(data, &h.spec); err != nil {
+		h.logger.Task().Warning(errors.Wrap(err, "unmarshalling test log spec; falling back to default spec"))
+		return
+	}
+
+	if err = h.spec.Format.validate(); err != nil {
+		h.logger.Task().Warning(errors.Wrapf(err, "invalid test log format specified; falling back to default text format"))
+	}
+}
+
+// followFile tails a test log file for the duration of a task, ingesting and
+// persisting log lines as they are written.
+func (h *testLogDirectoryHandler) followFile(ctx context.Context, path string) {
+	defer func() {
+		h.logger.Task().Critical(recovery.HandlePanicWithError(recover(), nil, "test log file follower"))
+	}()
+	defer h.wg.Done()
+
+	h.once.Do(h.getSpecFile)
+
+	h.logger.Task().Infof("new test log file '%s' found, initiating automated ingestion", path)
+
+	// The persisted log path should be relative to the reserved directory.
+	logPath, err := filepath.Rel(h.dir, path)
+	if err != nil {
+		h.logger.Task().Error(errors.Wrapf(err, "getting relative path for test log file '%s'", path))
+		return
+	}
+
+	t, err := tail.TailFile(path, tail.Config{
+		ReOpen:    true,
+		MustExist: true,
+		Follow:    true,
+		// Setting this field may lead to some data loss in the case of
+		// non-atomic line writes: if the last read line does not
+		// terminate with a newline character (e.g., the follower
+		// exits early or the test process crashes), it will not get
+		// returned. This is ok because the complete line must be read
+		// to guarantee proper line format for the parser (assumming
+		// the complete line has the expected format).
+		CompleteLines: true,
+	})
+	if err != nil {
+		h.logger.Task().Error(errors.Wrapf(err, "creating follower for test log file '%s'", path))
+		return
+	}
+	defer func() {
+		if err := t.Stop(); err != nil {
+			h.logger.Task().Error(errors.Wrapf(err, "following test log file '%s'", path))
+		}
+	}()
+
+	sender, err := h.createSender(ctx, logPath)
+	if err != nil {
+		h.logger.Task().Error(errors.Wrapf(err, "creating Sender for test log '%s'", path))
+		return
+	}
+	defer func() {
+		if err = t.Err(); err != nil {
+			h.logger.Task().Error(errors.Wrapf(err, "following test log file '%s'", path))
+		}
+		if err = sender.Close(); err != nil {
+			h.logger.Task().Error(errors.Wrapf(err, "closing Sender for test log '%s'", path))
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case line := <-t.Lines:
+			if line.Err != nil {
+				// If there is a line error it is safe to
+				// continue because the follower maintains the
+				// seek position and will pick up where it left
+				// off. These errors usually occur when rate
+				// limiting is configured and a cool down
```

Comment:
could you explain more what you mean by rate limiting?

Reply:
You can configure the file follower to rate limit the amount of data you read: https://pkg.go.dev/github.com/nxadm/tail#Config.

Reply:
Ah, got it. But do we need to check the err, then? (since we haven't enabled ratelimiting)

Reply:
Probably not, but I just added it to be safe and for posterity (in case we end up adding rate limiting).